src/iregnet_fit.cpp

```diff
@@ -538,3 +659,141 @@ compute_lambda_max(Rcpp::NumericMatrix X, double *w, double *z, double *eta,
+
+    ull none_censoring_number = 0;
+    ull left_or_right_censoring_number = 0;
+
+    /*
+     * Sort by censoring type,
+     */
+    for (ull i = 0; i < X.n_rows; ++i) {
+
+      if(status[i] == 1){
+
+        none_censoring_X_mat.row(none_censoring_number) = X.row(i);
+        none_censoring_y_mat.row(none_censoring_number) = y.row(i);
+
+        none_censoring_number++;
+      } else {
+
+        left_or_right_censoring_X_mat.row(left_or_right_censoring_number) = X.row(i);
+        left_or_right_censoring_y_mat.row(left_or_right_censoring_number) = y.row(i);
+
+        left_or_right_censoring_number++;
+      }
+    }
+
+    none_censoring_X_mat.resize(none_censoring_number, X.n_cols);
+    none_censoring_y_mat.resize(none_censoring_number, y.n_cols);
+
+    left_or_right_censoring_X_mat.resize(left_or_right_censoring_number, X.n_cols);
+    left_or_right_censoring_y_mat.resize(left_or_right_censoring_number, y.n_cols);
+
+    none_censoring_X_mat.insert_rows(none_censoring_number, left_or_right_censoring_X_mat);
+    none_censoring_y_mat.insert_rows(none_censoring_number, left_or_right_censoring_y_mat);
+
+    sorted_X = none_censoring_X_mat;
+    sorted_y = none_censoring_y_mat;
+
+    separator[0] = none_censoring_number;
+
+  } else {
+
+    /*
+     * Temp var for sort the matrix X&y
+     */
+    mat none_censoring_X_mat = mat(X.n_rows, X.n_cols);
+    mat left_censoring_X_mat = mat(X.n_rows, X.n_cols);
+    mat right_censoring_X_mat = mat(X.n_rows, X.n_cols);
+    mat interval_censoring_X_mat = mat(X.n_rows, X.n_cols);
+
+    mat none_censoring_y_mat = mat(y.n_rows, y.n_cols);
+    mat left_censoring_y_mat = mat(y.n_rows, y.n_cols);
+    mat right_censoring_y_mat = mat(y.n_rows, y.n_cols);
+    mat interval_censoring_y_mat = mat(y.n_rows, y.n_cols);
+
+    ull none_censoring_number = 0;
+    ull left_censoring_number = 0;
+    ull right_censoring_number = 0;
+    ull interval_censoring_number = 0;
+
+    /*
+     * Sort by censoring type,
+     */
+    for (ull i = 0; i < X.n_rows; ++i) {
+
+      if(status[i] == 1){
+
+        none_censoring_X_mat.row(none_censoring_number) = X.row(i);
+        none_censoring_y_mat.row(none_censoring_number) = y.row(i);
+
+        none_censoring_number++;
+      } else if(status[i] == 0){
+
+        right_censoring_X_mat.row(right_censoring_number) = X.row(i);
+        right_censoring_y_mat.row(right_censoring_number) = y.row(i);
+
+        right_censoring_number++;
+      } else if(status[i] == 2){
+
+        left_censoring_X_mat.row(left_censoring_number) = X.row(i);
+        left_censoring_y_mat.row(left_censoring_number) = y.row(i);
+
+        left_censoring_number++;
+      } else {
+
+        interval_censoring_X_mat.row(interval_censoring_number) = X.row(i);
+        interval_censoring_y_mat.row(interval_censoring_number) = y.row(i);
+
+        interval_censoring_number++;
+      }
+    }
+
+    none_censoring_X_mat.resize(none_censoring_number, X.n_cols);
```

Comment:
Is there a better way for concatenating different matrices?

Reply:
One thing that we could potentially do is to simply count the number of rows of each censoring type up front, create one new matrix, and keep inserting entries from the original matrix into the new matrix by maintaining pointers.